File: Unity/TSProj/src/Core/Module/Event/QueueMessageKit.ts
import { ActionType } from "../../../mods/ModTheSpire/Scripts/Action/AbstractGameAction";
import GainBlockAction from "../../../mods/ModTheSpire/Scripts/Action/Common/GainBlockAction";
import HealAction from "../../../mods/ModTheSpire/Scripts/Action/Common/HealAction";
import { UseCardAction } from "../../../mods/ModTheSpire/Scripts/Action/Utility/UseCardAction";
import {TimeKit} from "../../Utils/TimeKit";
import { IGameAction } from "./IGameAction";


export default class QueueMessageKit
{
    private waiting : boolean; 
    private list : Array<IGameAction> = new Array<IGameAction>();
    private currentRunning : IGameAction;
    public AddToBottom(func : IGameAction)
    {
        this.list.splice(0,0,func);
    }
    
    public AddToTop(func : IGameAction){
        this.list.push(func);
    }

    public Remove(func : IGameAction){
        this.list = this.list.filter(item => item == func);
    }

    public ClearPostCombatActions() {
        for (let index = this.list.length - 1; index >= 0 ; index--) {
            const e = this.list[index];
            if (e instanceof HealAction || e instanceof GainBlockAction || e instanceof UseCardAction || e.Type == ActionType.DAMAGE) continue;
            this.list.splice(index,1);
        }
    }

    public async Update()
    {
        if(this.waiting && this.currentRunning) {
            this.currentRunning += TimeKit.DeltaTime;
            return;
        }
        let action = this.currentRunning == undefined ? this.list.pop() : this.currentRunning;
        if(action){
            if(this.currentRunning != action){
                action.Init();
                this.currentRunning = action;
            }
            this.waiting = true;
            let result = await action.Update();
            this.waiting = false;
            if(result)
            {
                this.currentRunning = null;
            }
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Event/QueueMessageKit.cs
using System.Collections.Generic;
using Cysharp.Threading.Tasks;
using UnityEngine;

namespace ModTheSpire.Event
{
    public class QueueMessageKit
    {
        private static QueueMessageKit _inst;
        public static QueueMessageKit Inst => _inst ??= new QueueMessageKit();
        private QueueMessageKit()
        {
        }
        private bool Waiting { get; set; } 
        private readonly List<IGameAction> mList = new();
        private IGameAction CurrentRunning { get; set; }
        public void AddToBottom(IGameAction func)
        {
            mList.Insert(0,func);
        }
    
        public void AddToTop(IGameAction func){
            mList.Add(func);
        }

        public void Remove(IGameAction func){
            mList.RemoveAll(item => item == func);
        }

        public void ClearPostCombatActions()
        {
            for (var index = mList.Count - 1; index >= 0; index--)
            {
                var e = mList[index];
                //if (e is HealAction || e is GainBlockAction || e is UseCardAction || e.Type == ActionType.DAMAGE) continue;
                mList.RemoveAt(index);
            }
        }

        public async UniTask Update()
        {
            if(Waiting && CurrentRunning != null) {
                CurrentRunning.Duration += Time.deltaTime;
                return;
            }
            var action = this.CurrentRunning ?? this.mList[^1];
            if(action != null){
                if(this.CurrentRunning != action){
                    action.Init();
                    this.CurrentRunning = action;
                }
                this.Waiting = true;
                var result = await action.Update();
                this.Waiting = false;
                if(result)
                {
                    this.CurrentRunning = null;
                }
            }
        }
    }
}